crates/ecs/src/systems/mod.rs

```diff
@@ -1018,49 +1098,52 @@ macro_rules! impl_system_parameter_function {
                     <($([<P$parameter>],)*) as SystemParameters>::component_accesses()
                 }
 
-                fn iterates_over_entities() -> bool {
+                fn controls_iteration() -> bool {
                     false
                 }
 
                 fn base_signature() -> Option<TypeId> {
                     None
                 }
 
-                fn support_parallelization() -> bool {
+                fn supports_parallelization() -> bool {
                     <($([<P$parameter>],)*) as SystemParameters>::component_accesses()
                         .iter()
                         .all(|component_access| component_access.is_read())
                 }
             }
 
-            impl<'a, $([<P$parameter>]: SystemParameter,)*> IntoIterator for QuerySegment<'a, ($([<P$parameter>],)*)> {
+            impl<'a, $([<P$parameter>]: SystemParameter + 'a,)*> IntoIterator for QuerySegment<'a, ($([<P$parameter>],)*)> {
                 type Item = Query<'a, ($([<P$parameter>],)*)>;
                 type IntoIter = impl Iterator<Item=Self::Item>;
 
                 fn into_iter(mut self) -> Self::IntoIter {
                     iter::from_fn(move || {
                         Some(Query {
+                            // SAFETY: All Query instances need a mutable reference the segment for foreign entity access.
+                            // The components are only mutated when Write is a nested system parameter in Query.
+                            // In this case, intra-system-parallelization is disabled to maintain mutual exclusion.
```

Comment:
Same here, is there a fitting spot where we could add assertions that check this is always true? I.e. `assert!(query_has_no_write_parameters || !query_supports_intra_system_parallelization)`

Reply:
I don’t think there is a good way of checking this. The way you can see if a parameter has write access is to use component accesses and this is exactly what `supports_parallelization` does, so it would essentially be an asset for `supports_parallelization == supports_parallelization`.

Reply:
Oh, well. Then let's not!